### ISSUE TITLE
venv not updating re python version change

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-rpmbuild#29]
Thanks to @FroggyFlox and forum members Mark93 & eriklysoe we have a known failure re rpm update within our venv rebuild when we change our Python version. First, and currently most recently observed in rpm version-release 5.0.0-0.

Poetry, our dependency manager, has to date functioned flawlessly (even thought we use an early/old version) but we look now to have run into a limitation (old bug). During uninstall we remove .venv but during update we do not - intentionally. This enables re-use of existing unchanged, and so pre-build in a prior update/install, dependencies within our venv. However when we update our Python version, as was done fist in 5.0.0-0 (first release in our new testing phase) the required complete rebuild of the .venv, against the newly updted Python version, did not take place.

From @FroggyFlox's forum exposition here:
https://forum.rockstor.com/t/rockstor-pre-service-failing-after-update-to-5-0-0-0/8899/10
we have the following contents from our diagnostic file: poetry-install.txt, used to capture the output of our %posttrans scriptlet run of build.sh; which in turn invokes our run/rerun of a Poetry install - functioning as our venv maintenance re updating our venv:
```
cat poetry-install.txt 
Installing dependencies from lock file

  SolverProblemError

  The current project's Python requirement (2.7.18) is not compatible with some of the required packages Python requirement:
    - distro requires Python >=3.6, so it will not be satisfied for Python 2.7.18
  
  Because distro (1.8.0) requires Python >=3.6
   and no versions of distro match !=1.8.0, distro is forbidden.
  So, because rockstor depends on distro (*), version solving failed.

  at ~/.local/share/pypoetry/venv/lib64/python3.6/site-packages/poetry/puzzle/solver.py:241 in _solve
      237│             packages = result.packages
      238│         except OverrideNeeded as e:
      239│             return self.solve_in_compatibility_mode(e.overrides, use_latest=use_latest)
      240│         except SolveFailure as e:
    → 241│             raise SolverProblemError(e)
      242│ 
      243│         results = dict(
      244│             depth_first_search(
      245│                 PackageNode(self._package, packages), aggregate_package_nodes

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties
    
    For distro, a possible solution would be to set the `python` property to "<empty>"

    https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies,
    https://python-poetry.org/docs/dependency-specification/#using-environment-markers
```

**Comment by phillxnet** [comment]
It is proposed that we use the rpmbuild %prerun script under it's update identity only to perform a filesystem based wipe of our .venv: at least until we are using a more mature/up-to-date Poetry version in both our stable and testing branches: which in turn depends on the very updates we are looking at negotiation here (Python). This way we invoke a complete venv rebuild on every update.

We can later re-visit this approach with maybe a poetry based, rather than a filesystem based, approach to resetting our venv re Python versions. But for the time being we span quite the range of Python versions as we continue our major updates in the current testing phase.

**Comment by phillxnet** [comment]
Ordering of rpmbuild scriptlets execution during the various rpm operations:

https://docs.fedoraproject.org/en-US/packaging-guidelines/Scriptlets/#ordering

**Comment by phillxnet** [comment]
> It is proposed that we use the rpmbuild %preun script

Via some experimentation in draft pr #30 it was clarified that this rpm scriptlet is run from the package being updated, as such we cannot use it to deliver a fix in a single update.

So switch to the %post script that we already use for some directly clean-ups associated with our jslibs re-instantiation.

**Comment by phillxnet** [comment]
The %post script is run post install of the new version - but most importantly here, from the package being installed scriptlets. Hence a single update delivers the fix we require.

**Comment by phillxnet** [comment]
And given our build.sh is invoked after %post via %posttrans we are still enacting our required dir wipe before required.

**Comment by phillxnet** [comment]
With the caveat of introducing a command line reboot requirement post install (see: https://github.com/rockstor/rockstor-rpmbuild/issues/32)
Closing as:
Fixed by #31
